File: app/src/main/java/com/ignis/android_cleanarchitecture/data/repository/WeatherRepositoryImpl.java
package com.ignis.android_cleanarchitecture.data.repository;

import android.content.Context;

import com.ignis.android_cleanarchitecture.data.api.WeatherRetrofit;
import com.ignis.android_cleanarchitecture.data.api.WeatherApi;
import com.ignis.android_cleanarchitecture.data.dao.WeatherDao;
import com.ignis.android_cleanarchitecture.domain.model.WeatherModel;
import com.ignis.android_cleanarchitecture.domain.repository.WeatherRepository;

import io.realm.Realm;
import retrofit2.Retrofit;
import rx.Observable;

/**
 * Weather Repository Implement
 * <p>
 * Created by tamura_k on 2016/05/27.
 */
public class WeatherRepositoryImpl implements WeatherRepository {

    private Context context;

    public WeatherRepositoryImpl(Context context) {
        this.context = context;
    }

    @Override
    public Observable<WeatherModel> fetch(int cityId) {
        Retrofit retrofit = WeatherRetrofit.getInstance();
        WeatherApi weatherApi = retrofit.create(WeatherApi.class);
        return weatherApi.get(cityId);
    }

    @Override
    public WeatherModel find(Realm realm, int cityId) {
        WeatherDao weatherDao = new WeatherDao(realm);
        return weatherDao.find(cityId);
    }

    @Override
    public boolean exist(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            WeatherDao weatherDao = new WeatherDao(realm);
            return weatherDao.exist(cityId);
        } finally {
            if (realm != null) realm.close();
        }
    }

    @Override
    public void insert(WeatherModel weatherModel) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            realm.beginTransaction();
            WeatherDao weatherDao = new WeatherDao(realm);
            weatherDao.insert(weatherModel);
            realm.commitTransaction();
        } finally {
            if (realm != null) realm.close();
        }
    }

    @Override
    public void delete(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            realm.beginTransaction();
            WeatherDao weatherDao = new WeatherDao(realm);
            weatherDao.delete(cityId);
            realm.commitTransaction();
        } finally {
            if (realm != null) realm.close();
        }
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/data/repository/weather/WeatherRepositoryImpl.kt
package com.kazakago.cleanarchitecture.data.repository.weather

import android.content.Context
import com.kazakago.cleanarchitecture.data.database.dao.AppDatabase
import com.kazakago.cleanarchitecture.data.database.mapper.weather.WeatherEntityMapper
import com.kazakago.cleanarchitecture.domain.model.weather.Weather
import com.kazakago.cleanarchitecture.domain.repository.weather.WeatherRepository

class WeatherRepositoryImpl(private val context: Context) : WeatherRepository {

    private val database = AppDatabase.create(context)

    override fun find(cityId: String): Weather? {
        val weatherDao = database.weatherDao()
        val weather = weatherDao.findWeather(cityId)
        val location = weatherDao.findLocation(cityId)
        val description = weatherDao.findDescription(cityId)
        val forecasts = weatherDao.findForecasts(cityId)
        return WeatherEntityMapper.map(weather, location, description, forecasts)
    }

    override fun insert(weather: Weather) {
        val reverseMappingResult = WeatherEntityMapper.reverse(weather)
        val weatherDao = database.weatherDao()
        weatherDao.insert(reverseMappingResult.weatherEntity)
        weatherDao.insert(reverseMappingResult.locationEntity)
        weatherDao.insert(reverseMappingResult.descriptionEntity)
        weatherDao.insert(reverseMappingResult.forecastEntities)
    }

    override fun delete(weather: Weather) {
        val reverseMappingResult = WeatherEntityMapper.reverse(weather)
        val weatherDao = database.weatherDao()
        weatherDao.delete(reverseMappingResult.weatherEntity)
        weatherDao.delete(reverseMappingResult.locationEntity)
        weatherDao.delete(reverseMappingResult.descriptionEntity)
        weatherDao.delete(reverseMappingResult.forecastEntities)
    }

}
